### PR TITLE
Adds the ability to acquire readers in IndexShard

### PR DESCRIPTION
**Pull request by jimczi**
This change adds the ability to acquire a point in time reader on an engine.
This is needed for frozen indices that lazily loads the reader on every
phase of the search requests. 
Acquiring a reader on a frozen index ensures that the engine will not be closed until the reader is released, leaving the directory reader unopened until a call to acquire a searcher is made.
When the searcher is closed, the underlyinng directory reader is also closed unless another requests on the same frozen shard is in-flight. 
This ensures that the directory reader of frozen indices is opened only when requests are
executed (they consume a thread in the search throttled pool).

### CODE REVIEW

**Review by dnhatn — COMMENTED** [review]
Thanks, Jim! I like the fact that we have simplified SearchOperationListener significantly in this change. However, I wonder if we can avoid introducing a new wrapper (i.e., Engine#Reader). Can we introduce unload/reload methods to `Engine. Searcher`, then call them in ReaderContext and LegacyReaderContext instead? I might be missing something here.

**Comment by jimczi** [comment]
I pushed a change to limit reader contexts to the user that created it. @dnhatn can you take another look ?

File: server/src/main/java/org/elasticsearch/action/OriginalIndices.java [hunk 1]
```diff
@@ -57,13 +58,20 @@ public IndicesOptions indicesOptions() {
     }
 
     public static OriginalIndices readOriginalIndices(StreamInput in) throws IOException {
+        if (in.getVersion().onOrAfter(Version.V_8_0_0) && in.readBoolean() == false) {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/action/OriginalIndices.java:61]: Should we move the logic here to SearchShardTarget instead?

File: test/framework/src/main/java/org/elasticsearch/test/engine/MockInternalEngine.java [hunk 2]
```diff
@@ -81,4 +82,20 @@ private void flushAndCloseInternal() throws IOException {
         final Engine.Searcher engineSearcher = super.acquireSearcher(source, scope);
         return support().wrapSearcher(engineSearcher);
     }
+
+    @Override
+    public SearcherSupplier acquireSearcherSupplier(Function<Searcher, Searcher> wrapper, SearcherScope scope) throws EngineException {
+        final SearcherSupplier reader = super.acquireSearcherSupplier(wrapper, scope);
```
Comment (dnhatn) [inline test/framework/src/main/java/org/elasticsearch/test/engine/MockInternalEngine.java:88]: Maybe `return super.acquireSearcherSupplier(wrapper.andThen(support::wrapSearcher), scope)`?

File: x-pack/plugin/frozen-indices/src/main/java/org/elasticsearch/xpack/frozen/FrozenIndices.java [hunk 3]
```diff
@@ -55,14 +54,6 @@
         return Arrays.asList(FrozenEngine.INDEX_FROZEN);
     }
 
-    @Override
-    public void onIndexModule(IndexModule indexModule) {
-        if (FrozenEngine.INDEX_FROZEN.get(indexModule.getSettings())) {
-            indexModule.addSearchOperationListener(new FrozenEngine.ReacquireEngineSearcherListener());
```
Comment (dnhatn) [inline x-pack/plugin/frozen-indices/src/main/java/org/elasticsearch/xpack/frozen/FrozenIndices.java:61]: 💯

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/SecuritySearchOperationListener.java [hunk 4]
```diff
@@ -44,25 +45,32 @@ public SecuritySearchOperationListener(SecurityContext securityContext, XPackLic
...
+     * compares the {@link Authentication} object from the reader context with the current
+     * authentication context
      */
     @Override
-    public void validateSearchContext(SearchContext searchContext, TransportRequest request) {
+    public void validateSearchContext(ReaderContext readerContext, SearchContext searchContext, TransportRequest request) {
         if (licenseState.isSecurityEnabled()) {
             if (searchContext.scrollContext() != null) {
-                final Authentication originalAuth = searchContext.scrollContext().getFromContext(AuthenticationField.AUTHENTICATION_KEY);
+                final Authentication originalAuth = readerContext.getFromContext(AuthenticationField.AUTHENTICATION_KEY);
```
Comment (dnhatn) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/SecuritySearchOperationListener.java:73]: I think we should validate the authentication of any reader context, not only the reader contexts with scroll?

**Review by dnhatn — COMMENTED** [review]
I've left some comments, but this looks great. Thanks Jim!

File: server/src/main/java/org/elasticsearch/search/SearchService.java [hunk 5]
```diff
@@ -414,20 +423,16 @@ public void onFailure(Exception exc) {
...
-                try {
-                    result = command.get();
-                } catch (Exception exc) {
-                    listener.onFailure(exc);
-                    return;
-                }
-                listener.onResponse(result);
-            });
-        } catch (Exception exc) {
-            listener.onFailure(exc);
```
[the remark this answers is not in the document]
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:429]: We handle exception when calling from query and dfs. I removed the usage of the search thread pool when creating a reader (openReaderContext), I think it's ok to open in the listener thread pool.

File: server/src/main/java/org/elasticsearch/action/OriginalIndices.java (hunk 1 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/action/OriginalIndices.java]: sorry it's not needed. I pushed a change to cleanup the usage of search shard target when opening a new reader:
https://github.com/elastic/elasticsearch/pull/54966/commits/0ec84abf64beeae6c5dbb35217306e447ca725d8

File: test/framework/src/main/java/org/elasticsearch/test/engine/MockInternalEngine.java (hunk 2 above)
Comment (jimczi) [reply test/framework/src/main/java/org/elasticsearch/test/engine/MockInternalEngine.java]: ++, I pushed https://github.com/elastic/elasticsearch/pull/54966/commits/0ec84abf64beeae6c5dbb35217306e447ca725d8

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/SecuritySearchOperationListener.java (hunk 4 above)
Comment (jimczi) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/SecuritySearchOperationListener.java]: Oups, I pushed a change to address this:
https://github.com/elastic/elasticsearch/pull/54966/commits/0ec84abf64beeae6c5dbb35217306e447ca725d8

**Review by dnhatn — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/search/SearchService.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/search/SearchService.java:429]: Looks good. Thanks for the explanation.

**Comment by jimczi** [comment]
Thanks @dnhatn

**Pull request merged**